bonsai/BudgetTransactions/BudgetTransactions.swift

```diff
@@ -0,0 +1,175 @@
+//
+//  BudgetTransactions.swift
+//  bonsai
+//
+//  Created by hoang on 08.12.2021.
+//
+
+import SwiftUI
+
+struct VisualEffectView: UIViewRepresentable {
+    var effect: UIVisualEffect?
+    func makeUIView(context: UIViewRepresentableContext<Self>) -> UIVisualEffectView { UIVisualEffectView() }
+    func updateUIView(_ uiView: UIVisualEffectView, context: UIViewRepresentableContext<Self>) { uiView.effect = effect }
+}
+
+struct RoundedCorner: Shape {
+
+    var radius: CGFloat = .infinity
+    var corners: UIRectCorner = .allCorners
+
+    func path(in rect: CGRect) -> Path {
+        let path = UIBezierPath(roundedRect: rect, byRoundingCorners: corners, cornerRadii: CGSize(width: radius, height: radius))
+        return Path(path.cgPath)
+    }
+}
+
+extension View {
+    func cornerRadius(_ radius: CGFloat, corners: UIRectCorner) -> some View {
+        clipShape(RoundedCorner(radius: radius, corners: corners))
+    }
+}
+
+
+struct BudgetTransactions: View {
+    
+    private typealias TransactionGroup = [String: [Transaction]]
+    private let transactionsByDate: TransactionGroup
+    
+    init(transactions: [Transaction]) {
+        var dict: TransactionGroup = [:]
+        for transaction in transactions {
+            let date = Date.dateMapper(date: transaction.date)
+            if dict.keys.contains(date) {
+                if var arr = dict[date] {
+                    arr.append(transaction)
+                    dict[date] = arr
+                }
+            } else {
+                dict[date] = [transaction]
+            }
+        }
```

Comment:
Still O(n^2) :(

you iterate over all transactions once
and you iterate over all or some of the keys (contains)
you actually don't need to check for `contains` here, because [key] access is safe (returns Optional)

You could take advantage of dictionary key subscript access to make it O(n)

```swift
for transaction in transactions {
  let date = transaction.date.mappedToString
  if var arr = dict[date] {
    arr.append(transaction)
    dict[date] = arr
  } else {
    dict[date] = [transaction]
  }
}
```

Reply:
I think sorting also should be done here using swift-collections `SortedDictionary`

Reply:
added  OrderedDictionary

Reply:
Also my thoughts in Future.  
We can sort by date when we make fetch results by predicate and send next to unnessesary screens with sorted transactions, actually i dont see any case where we used unsorted transactions yet :)